File: lib/models/search_params.dart
import 'package:collection/collection.dart';
import 'package:equatable/equatable.dart';

part 'search_filter.dart';

class SearchParams extends Equatable {
  const SearchParams({
    required this.filters,
    required this.query,
    required this.page,
    required this.sorted,
    required this.exactMatch,
  });

  SearchParams.init()
      : filters = <SearchFilter>{},
        query = '',
        page = 0,
        sorted = false,
        exactMatch = false;

  final Set<SearchFilter> filters;
  final String query;
  final int page;
  final bool sorted;
  final bool exactMatch;

  SearchParams copyWith({
    Set<SearchFilter>? filters,
    String? query,
    int? page,
    bool? sorted,
    bool? exactMatch,
  }) {
    return SearchParams(
      filters: filters ?? this.filters,
      query: query ?? this.query,
      page: page ?? this.page,
      sorted: sorted ?? this.sorted,
      exactMatch: exactMatch ?? this.exactMatch,
    );
  }

  SearchParams copyWithFilterRemoved<T extends SearchFilter>() {
    return SearchParams(
      filters: <SearchFilter>{...filters}
        ..removeWhere((SearchFilter e) => e is T),
      query: query,
      page: page,
      sorted: sorted,
      exactMatch: exactMatch,
    );
  }

  SearchParams copyWithFilterAdded(
    SearchFilter filter,
  ) {
    return SearchParams(
      filters: <SearchFilter>{...filters, filter},
      query: query,
      page: page,
      sorted: sorted,
      exactMatch: exactMatch,
    );
  }

  String get filteredQuery {
    final StringBuffer buffer = StringBuffer();
    final String encodedQuery =
        Uri.encodeComponent(exactMatch ? '"$query"' : query);

    if (sorted) {
      buffer.write('search_by_date?query=$encodedQuery');
    } else {
      buffer.write('search?query=$encodedQuery');
    }

    final Iterable<NumericFilter> numericFilters =
        filters.whereType<NumericFilter>();
    final List<TagFilter> tagFilters = <TagFilter>[
      ...filters.whereType<TagFilter>(),
    ];

    if (numericFilters.isNotEmpty) {
      buffer
        ..write('&numericFilters=')
        ..writeAll(
          numericFilters.map<String>((NumericFilter e) => e.query),
          ',',
        );
    }

    if (tagFilters.isNotEmpty) {
      buffer
        ..write('&tags=')
        ..writeAll(
          tagFilters.map<String>((TagFilter e) => e.query),
          ',',
        );
    }

    buffer.write('&page=$page');

    return buffer.toString();
  }

  bool contains<T extends SearchFilter>() {
    return filters.whereType<T>().isNotEmpty;
  }

  T? get<T extends SearchFilter>() {
    return filters.singleWhereOrNull(
      (SearchFilter e) => e is T,
    ) as T?;
  }

  @override
  List<Object?> get props => <Object?>[
        filters,
        query,
        page,
        sorted,
        exactMatch,
      ];
}

File: ios/StoryWidget/Models/SearchParams.swift
import Foundation

public class SearchParams: Equatable {
    public let page: Int
    public let query: String
    public let sorted: Bool
    public let filters: Set<SearchFilter>
    
    public var filteredQuery: String {
        var buffer = String()
        
        if sorted {
            buffer.append("search_by_date?query=\(query)")
        } else {
            buffer.append("search?query=\(query)")
        }
        
        if !filters.isEmpty {
            let numericFilters = filters.filter({ $0.isNumericFilter })
            let tagFilters = filters.filter({ $0.isTagFilter })
            
            if !numericFilters.isEmpty {
                buffer.append("&numericFilters=")
                for filter in filters.filter({ $0.isNumericFilter }) {
                    buffer.append(filter.query)
                    buffer.append(",")
                }
                buffer = String(buffer.dropLast(1))
            }

            if !tagFilters.isEmpty {
                buffer.append("&tags=(")
                for filter in filters.filter({ $0.isTagFilter }) {
                    buffer.append(filter.query)
                    buffer.append(",")
                }
                buffer = String(buffer.dropLast(1))
                buffer.append(")")
            }

        }
        
        buffer.append("&page=\(page)");
        
        return buffer
    }
    
    public init(page: Int, query: String, sorted: Bool, filters: Set<SearchFilter>) {
        self.page = page
        self.query = query
        self.sorted = sorted
        self.filters = filters
    }
    
    public func copyWith(page: Int? = nil, query: String? = nil, sorted: Bool? = nil, filters: Set<SearchFilter>? = nil) -> SearchParams {
        return SearchParams(page: page ?? self.page, query: query ?? self.query, sorted: sorted ?? self.sorted, filters: filters ?? self.filters)
    }
    
    public static func == (lhs: SearchParams, rhs: SearchParams) -> Bool {
        return lhs.page == rhs.page && lhs.query == rhs.query && lhs.sorted == rhs.sorted && lhs.filters == rhs.filters
    }
}
